include/libtorrent/alert_types.hpp

```diff
@@ -3116,6 +3118,27 @@ TORRENT_VERSION_NAMESPACE_3_END
 	TORRENT_EXTRA_EXPORT char const* performance_warning_str(performance_alert::performance_warning_t i);
 
 
+	template<typename T>
+	class alert_exception : public std::exception {
+	public:
+		alert_exception(const T* a) : m_alert(a) {}
+
+		const char* what() const noexcept override {
+			return m_alert->message().c_str();
+		}
+
+	private:
+		const T* m_alert;
```

Comment:
this looks a bit suspicious. alert allocated in the heterogeneous queue have their lifetimes tied to that queue. throwing exceptions with pointers into it seems like asking for accessing danging pointers